R/calc_coherence.R

```diff
@@ -0,0 +1,259 @@
+  #e.g. by defining them as reference/S6 class, this would also allow more flexible definition of additional measures by user
+  coh_funs <- list(
+    #LOG-RATIO
+    logratio_UMass = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant) {log(log_smooth_constant + tcm[wi,wj]) - log(log_smooth_constant + tcm[wj,wj])}
+                         ,set_type = "one_pre_topic_order"
+                         ,aggr_fun = "function(x) sum(x, na.rm = T)")
+    #smoothing parameter = 1 resembles UMAss, .01 resembles stm package, default as in paper by Röder, i.e. .1e-12
+    ,logratio = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant) {log(log_smooth_constant + tcm[wi,wj]) - log(log_smooth_constant + tcm[wj,wj])}
+                         ,set_type = "one_pre"
+                         ,aggr_fun = "function(x) sum(x, na.rm = T)")
+    ,prob_logratio = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant)  {log(log_smooth_constant + (tcm[wi,wj]/ndocs_tcm)) - log(tcm[wj,wj]/ndocs_tcm)}
+                               ,set_type = "one_pre"
+                               ,aggr_fun = "function(x) mean(x, na.rm = T)")
+    #PMI
+    #format of PMI formula proposed by @andland - https://github.com/dselivanov/text2vec/issues/236
+    ,pmi = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant)  {log2((tcm[wi,wj]/ndocs_tcm) + log_smooth_constant) - log2(tcm[wi,wi]/ndocs_tcm) - log2(tcm[wj,wj]/ndocs_tcm)}
+                     ,set_type = "one_pre"
+                     ,aggr_fun = "function(x) mean(x, na.rm = T)")
+    #NORMALIZED PMI
+    #again, in contrast, to other implementations, only intrinsic NPMI as in PMIM (for implementation with sliding window see, e.g., Bouma, 2009)
+    ,npmi = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant) {(log2((tcm[wi,wj]/ndocs_tcm) + log_smooth_constant) - log2(tcm[wi,wi]/ndocs_tcm) - log2(tcm[wj,wj]/ndocs_tcm)) /  -log2((tcm[wi,wj]/ndocs_tcm) + log_smooth_constant)}
+                      ,set_type = "one_pre"
+                      ,aggr_fun = "function(x) mean(x, na.rm = T)")
+    #DIFFERENCE
+    #assuming we use ordered tcm it follows that p(wi)>p(wj)
+    #to set bounds of the measures [-1,1] (1 is good)  wi/wj are switched in formula
+    #this is similar (not exactly the same) to the measure of textmineR package https://github.com/TommyJones/textmineR/issues/35
+    ,prob_dif = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant) {tcm[wj,wi]/tcm[wj,wj] - (tcm[wj,wj]/ndocs_tcm)}
+                          ,set_type = "one_suc"
+                          ,aggr_fun = "function(x) mean(x, na.rm = T)")
+  )
+
+#WRAPPER FUNCTION TO CALCULATE COHERENCE-------------------------------------------------------
+  #TODO coh_funs and other args need to be passed to the function as argument,
+  #later steps would be less verbose when fetching several or the arguments,e.g., tcm, from the function environment -> scoping
+  calc_coh <- function( idxs
```

Comment:
quite complex logic here and in for loop below. I would just assign different coherence measures line by line (more or less copy-paste). Later we can think whether it worth to generalize it to user-defined coherence measures.